### PR TITLE
add getDDOFromEthereum

### PR DESCRIPTION
**Pull request by xinbilifeid**

### CODE REVIEW

File: src/services/ddo.ts [hunk 1]
```diff
@@ -16,7 +16,32 @@ function getDDODetails(did: string): DDODetails {
 // This function expects at least one public key
 // TODO: Add created and update at fields
 function generateDDO(ddoDetails: DDODetails): DDO {
-  return _generateBaseDDO(ddoDetails);
+  if (ddoDetails.created != null) return _generateDDO(ddoDetails);
+  else return _generateBaseDDO(ddoDetails);
```
Comment (jonohrt) [inline src/services/ddo.ts]: No need for this `else`.

File: src/services/ddo.ts [hunk 2]
```diff
@@ -16,7 +16,32 @@ function getDDODetails(did: string): DDODetails {
 // This function expects at least one public key
 // TODO: Add created and update at fields
 function generateDDO(ddoDetails: DDODetails): DDO {
-  return _generateBaseDDO(ddoDetails);
+  if (ddoDetails.created != null) return _generateDDO(ddoDetails);
```
Comment (jonohrt) [inline src/services/ddo.ts]: A bunch of things here, first, as a javascript-ism... never use  == or !=. Always use === or !==. This should have been caught by the linter. Are you checking the lint messages? As soon as CI is setup, this won't pass. 
I prefer to always use curly brackets on if statements as a style choice.  I think it's easier to read. 
In javascript checking for null is unsafe.  null !== undefined and could lead to unexpected errors. A safer version of this is `if(!ddoDetails.created) {}` or you could refactor this whole function to the ternary operator version ` return ddoDetails.created ? _generateDDO(ddoDetails) :  _generatebaseDDO(ddoDetails)`  OR.... better yet, see my comment 2 down

File: src/services/ddo.ts [hunk 3]
```diff
@@ -16,7 +16,32 @@ function getDDODetails(did: string): DDODetails {
...
+  else return _generateBaseDDO(ddoDetails);
+}
+
```
Comment (jonohrt) [inline src/services/ddo.ts]: The intention for _generateDDO is that it should inherit off of the _generateBaseDDO function below it. My intention was for both offline and online DDOs should run through the same _generateDDO function once it was created.

File: src/index.ts [hunk 4]
```diff
@@ -52,15 +60,15 @@ function validateDID(did: any | string): Promise<boolean> {
...
     .then(getIDString)
-    .then(getKey)
+    .then(getKey);
 }
 
 function getNetowrkFromDID(did: string): Promise<string> {
   return Promise.resolve(did)
-  .then(getIDString)
-  .then(getNetwork)
+    .then(getIDString)
```
Comment (jonohrt) [inline src/index.ts]: Can you check that VSCode is set to `spaces: 2` in the bottom right bar?  It looks like the tabs are being changed.

**Comment by xinbilifeid** [comment]
https://lifeid.atlassian.net/browse/LV-138

File: src/services/ddo.ts [hunk 5]
```diff
@@ -15,15 +17,21 @@ function getDDODetails(did: string): DDODetails {
...
-function _generateBaseDDO(ddoDetails: DDODetails) {
+function _generateBaseDDO(ddoDetails: DDODetails): DDO {
   return {
     "@context": "https://w3id.org/did/v1",
     id: ddoDetails.did,
+    //name: ddoDetails.name,
     entityType: EntityTypes[ddoDetails.entityType],
+    //created: ddoDetails.created,
+    //updated: ddoDetails.updated,
     publicKeys: [
```
Comment (jonohrt) [inline src/services/ddo.ts]: Please remove commented code

File: src/services/ddo.ts [hunk 6]
```diff
@@ -15,15 +17,21 @@ function getDDODetails(did: string): DDODetails {
 }
 // This function expects at least one public key
 // TODO: Add created and update at fields
-function generateDDO(ddoDetails: DDODetails): DDO {
-  return _generateBaseDDO(ddoDetails);
+function generateDDO(ddoDetails: DDODetails): Promise<DDO> {
+  return Promise.resolve(ddoDetails)
+    .then(_generateBaseDDO)
+    .then(ddo => _addName(ddo, ddoDetails.name))
+    .then(ddo => _addDates(ddo, ddoDetails.created, ddoDetails.updated));
```
Comment (jonohrt) [inline src/services/ddo.ts]: This is way better!

File: test/services/ddo.test.ts [hunk 7]
```diff
@@ -18,136 +18,147 @@ describe("ddo.ts", () => {
...
         } else {
           throw new Error("None shall pass!");
         }
-      } else {
-        throw new Error("None shall pass!");
-      }
+      });
     });
   });
   describe("getDDODetails", () => {
```
Comment (jonohrt) [inline test/services/ddo.test.ts]: Can you add tests to make sure your dates and name is being generated on the DDO? This library is going to be open sourced and used by a bunch of different software so I want to make sure we keep up with tests.

File: src/services/ddo.ts [hunk 8]
```diff
@@ -55,8 +52,8 @@ function _addName(ddo: DDO, name?: string): DDO {
 }
 
 function _addDates(ddo: DDO, created?: string, updated?: string): DDO {
-  if (created) ddo.created = format(created, TIMESTAMP_FORMAT);
-  if (updated) ddo.updated = format(updated, TIMESTAMP_FORMAT);
+  if (created) ddo.created = format(parseInt(created, 10), TIMESTAMP_FORMAT);
+  if (updated) ddo.updated = format(parseInt(updated, 10), TIMESTAMP_FORMAT);
```
Comment (jonohrt) [inline src/services/ddo.ts]: Do we want to make a util function for `format(parseInt(created, 10), TIMESTAMP_FORMAT);`?

File: test/services/ddo.test.ts [hunk 9]
```diff
@@ -29,12 +32,30 @@ describe("ddo.ts", () => {
...
+    });
+    it("should generate a valid updated", () => {
+      generateDDO(ddoDetails).then(result => {
+        result.should.have.property("name");
+        if (result.updated) result.updated.should.equal("1970-01-18T08:40:38Z");
+      });
+    });
     it("should have public keys", () => {
       generateDDO(ddoDetails).then(result => {
         result.should.have.property("publicKeys");
```
Comment (jonohrt) [inline test/services/ddo.test.ts]: You might want to write a test to make sure the DDO doesn't contain the optional fields if they are missing from the details.

**Pull request merged**